internal/jennies/java/converter.go

```diff
@@ -0,0 +1,167 @@
+	"github.com/grafana/cog/internal/orderedmap"
+	"github.com/grafana/cog/internal/tools"
+)
+
+const variableModel = "VariableModel"
+
+type Converter struct {
+	config         Config
+	nullableConfig languages.NullableConfig
+	tmpl           *template.Template
+}
+
+func (jenny *Converter) JennyName() string {
+	return "JavaConverter"
+}
+
+func (jenny *Converter) Generate(context languages.Context) (codejen.Files, error) {
+	files := codejen.Files{}
+
+	for _, builder := range context.Builders {
+		output, err := jenny.generateConverter(context, builder)
+		if err != nil {
+			return nil, err
+		}
+
+		filename := filepath.Join(
+			jenny.config.ProjectPath,
+			formatPackageName(builder.Package),
+			fmt.Sprintf("%sConverter.java", tools.UpperCamelCase(builder.Name)),
+		)
+
+		files = append(files, *codejen.NewFile(filename, output, jenny))
+	}
+
+	var err error
+	for _, schema := range context.Schemas {
+		filterDataqueryObjects(schema).Iterate(func(key string, obj ast.Object) {
+			output, genErr := jenny.generateDataqueryConverter(context, schema, obj)
+			if genErr != nil {
+				err = genErr
+			} else {
+				filename := filepath.Join(
+					jenny.config.ProjectPath,
+					formatPackageName(schema.Package),
+					fmt.Sprintf("%sMapperConverter.java", tools.UpperCamelCase(obj.Name)),
+				)
+
+				files = append(files, *codejen.NewFile(filename, output, jenny))
+			}
+		})
+	}
+
+	return files, err
+}
+
+func filterDataqueryObjects(schema *ast.Schema) *orderedmap.Map[string, ast.Object] {
+	filteredObjects := orderedmap.New[string, ast.Object]()
+
+	schema.Objects.Iterate(func(key string, obj ast.Object) {
+		if obj.Type.ImplementedVariant() == string(ast.SchemaVariantDataQuery) {
+			if !obj.Type.HasHint(ast.HintSkipVariantPluginRegistration) {
+				filteredObjects.Set(key, obj)
+			}
+		}
+	})
+
+	return filteredObjects
+}
+
+func (jenny *Converter) generateConverter(context languages.Context, builder ast.Builder) ([]byte, error) {
+	converter := languages.NewConverterGenerator(jenny.nullableConfig).FromBuilder(context, builder)
+
+	schema, schemaFound := context.Schemas.Locate(builder.Package)
+	isPanel := schemaFound && schema.Metadata.Variant == ast.SchemaVariantPanel && builder.Name == "Panel"
+
+	imports := NewImportMap(jenny.config.PackagePath)
+	packageMapper := func(pkg string, class string) string {
+		if imports.IsIdentical(pkg, schema.Package) {
+			return ""
+		}
+
+		return imports.Add(class, pkg)
+	}
+	typeFormatter := createFormatter(context, jenny.config).withPackageMapper(packageMapper)
+
+	builderNameFormat := fmt.Sprintf("%s.Builder", tools.UpperCamelCase(converter.BuilderName))
+	if isPanel {
+		builderNameFormat = fmt.Sprintf("%s.PanelBuilder", typeFormatter.formatPackage(converter.Package))
+	} else if converter.BuilderName != converter.Input.TypeRef.ReferredType {
+		if converter.Input.TypeRef.ReferredType == variableModel { // TODO: Find a better way to identify this...
```

Comment:
Agreed, let's look at this together: this is going to break with the Dashboard v2 schema :/

also: ideally the jennies should know nothing about the schemas they're generating code for